### ISSUE TITLE
Pre-LN&Post-LN 

### ISSUE THREAD

**Issue by xyhe1996** [ModelTC/Outlier_Suppression_Plus#5]
https://github.com/ModelTC/Outlier_Suppression_Plus/blob/3ba97ae2dab0e6e5ead5da1795f50fd47025a49d/quant_transformer/model/quant_llama.py#L242  hi~ i noticed that there are two types LN i.e. pre-LN&post are shown in Fig2. while it seems like that only the pre-LN is used in the above code. if i missed some details, please let me know

**Comment by wimh966** [comment]
yeah, you are right. These days, big models usually take the pre-LN structure as it can make training more stable. And BERT takes the Post-LN structure.

**Comment by xyhe1996** [comment]
> yeah, you are right. These days, big models usually take the pre-LN structure as it can make training more stable. And BERT takes the Post-LN structure.

appreciate your timely reply. one more question: how to understand the channel for the activations with shape [num_batches, num_seqs/tokens, num_embedding] in your paper? weather the channel represents the num_seqs/tokens dim or not?  If yes, then how to figure out the per_channel and per_token in paper?

**Comment by wimh966** [comment]
The channel means the embedding or the hidden state dimension. Per-channel means taking different quantization parameters for different channels or hidden states dimensions. Per-channel for weight is allowed but for activation is not as it will destroy the integer matrix multiplication. Per-token means taking different quantization parameters for different tokens.

**Comment by xyhe1996** [comment]
> The channel means the embedding or the hidden state dimension. Per-channel means taking different quantization parameters for different channels or hidden states dimensions. Per-channel for weight is allowed but for activation is not as it will destroy the integer matrix multiplication. Per-token means taking different quantization parameters for different tokens.

thanks. so for activations, the Per-token quantization is employed in this paper, which means all num_embedding values for one specific token of num_tokens dim share identical quantization params.  am i right?

**Comment by wimh966** [comment]
Yes, you are right.

**Comment by xyhe1996** [comment]
> The channel means the embedding or the hidden state dimension. Per-channel means taking different quantization parameters for different channels or hidden states dimensions. Per-channel for weight is allowed but for activation is not as it will destroy the integer matrix multiplication. Per-token means taking different quantization parameters for different tokens.

thx again, could you further explain why the Per-channel quantization for activations will destroy the integer matrix multiplication? what does the integer matrix multiplication refer to?

**Comment by wimh966** [comment]
I think you can refer to this paper here. Quantization and training of neural networks for efficient integer-arithmetic-only inference

**Comment by xyhe1996** [comment]
> I think you can refer to this paper here. Quantization and training of neural networks for efficient integer-arithmetic-only inference

thanks a million~ in the reference, i find that the reason why the integer matrix multiplication will be destroyed is the multiplier $M:=\frac{S_1S_2}{S_3}$ can not be distilled in  formula $(3)$ when the Per_channel quantization for activations is employed.  Did I get it right?

**Comment by wimh966** [comment]
Yes, you're right.

**Comment by xyhe1996** [comment]
> Yes, you're right.

thx for your help~

**Comment by flymmmfly** [comment]
> The channel means the embedding or the hidden state dimension. Per-channel means taking different quantization parameters for different channels or hidden states dimensions. Per-channel for weight is allowed but for activation is not as it will destroy the integer matrix multiplication. Per-token means taking different quantization parameters for different tokens.

Hello, I have a question. For example, assuming the dimension of the input activation is (batchsize, 512, 4096) and the dimension of the weight is (4096, 1024), So, for the two dimensions of the weight, along which dimension should we perform per-channel quantization? thanks

**Comment by wimh966** [comment]
Hello, it's the output dimension that won't destroy the inner multiplication and addition. I assume in your case, the output dimension of weight would be 1024?

**Comment by flymmmfly** [comment]
> Hello, it's the output dimension that won't destroy the inner multiplication and addition. I assume in your case, the output dimension of weight would be 1024?

Yes, the output dimension is 1024! I am very happy to receive your reply, but when I read a lot of algorithm codes, such as AWQ, I found that they grouped and quantized the weights according to the input dimension, which confused me...
https://github.com/mit-han-lab/llm-awq/blob/41d7e704932b7eaf8b3dcfb21e1fe49fd296d894/awq/quantize/auto_clip.py#L16

**Comment by wimh966** [comment]
Normally, we write the weight in the way (output_dimension, input_dimension). I think maybe we have the different understanding of input and output dimensions.

**Comment by flymmmfly** [comment]
Well, I understand that weights in pytorch are usually expressed as (output_dimension, input_dimension), and the weights need to be transposed when performing matrix multiplication. If we go back to the example in AWQ below, 
`# w           [co, ci]      -> [co, 1, n_group, group size]`
co=1024, ci=4096, then we should call ci the input dimension, right? If ci is the input dimension, then here the input dimension is grouped and quantized, which will affect the inner multiplication and addition. Am I right?

**Comment by wimh966** [comment]
For the AWQ question, maybe you can ask them directly via the github issue? Because I'm not the author of that and didn't try that yet, I'm not so clear about what they do. For this repo, I think we're doing on the output dimension.

**Comment by flymmmfly** [comment]
Thanks！

**Comment by wimh966** [comment]
BTW, if it's group quantization for weights, I think it's normal that people do on the input dimension with limited groups.

**Comment by flymmmfly** [comment]
Alright, thanks for your patient reply~
